### ISSUE TITLE
authorizedCallback called before wellknown endpoint is discovered

### ISSUE THREAD

**Issue by Silthus** [damienbod/angular-auth-oidc-client#50]
The authorize callback function tries to access the jwk_uris from the well known discovery endpoint without checking and waiting for the endpoint to load.

In certain edge cases, where the configuring module is loaded later, and the discovery of the well known configuration arrives after the call of the authorizeCallback, the authorizeCallback fails with the following error message:

> jwks_uri: undefined
> angular-auth-oidc-client.es5.js:1495 Unexpected token < in JSON at position 0
> OidcSecurityService.handleErrorGetSigningKeys @ angular-auth-oidc-client.es5.js:1495
> CatchSubscriber.error @ catch.js:104
> MapSubscriber._next @ map.js:80
> Subscriber.next @ Subscriber.js:89
> onLoad @ http.es5.js:1226
> ZoneDelegate.invokeTask @ zone.js:424
> onInvokeTask @ core.es5.js:3924
> ZoneDelegate.invokeTask @ zone.js:423
> Zone.runTask @ zone.js:191
> ZoneTask.invoke @ zone.js:486
> core.es5.js:1020 ERROR Unexpected token < in JSON at position 0

To fix such timing issues it would be best to wait for the well known config Observable to complete and then parse the callback token.

```
this.authWellKnownEndpoints.onWellKnownEndpointsLoaded
            .concat(this.http.get(this.authWellKnownEndpoints.jwks_uri)
            .map(this.extractData)
            .catch(this.handleErrorGetSigningKeys));
```

**Comment by damienbod** [comment]
thanks for this.

**Comment by craigpayne** [comment]
Experiencing the same problem.  Has this been fixed? Authorize firing before discovery endpoint  loaded

**Comment by damienbod** [comment]
@craigpayne  no, plan to do this as soon as I have access to a PC, in 2 weeks

**Comment by damienbod** [comment]
You could also subscribe to the onModuleSetup event before calling the authorize or authorizeCallback functions see the api docs for an example

**Comment by damienbod** [comment]
@Silthus @craigpayne Is the following solution ok for you:

subscribe to the onModuleSetup event:

```
 constructor(public oidcSecurityService: OidcSecurityService) {
        this.oidcSecurityService.onModuleSetup.subscribe(() => { this.onModuleSetup(); });
    }
```

Handle the authorize callback using the event:
```
 private onModuleSetup() {
        if (window.location.hash) {
            this.oidcSecurityService.authorizedCallback();
        }
    }
```

unsubscribe
```
 ngOnDestroy(): void {
        this.oidcSecurityService.onModuleSetup.unsubscribe();
    }
```

**Comment by Silthus** [comment]
Looks great 👍🏻

**Comment by damienbod** [comment]
just testing, this doesn't always work because the onModuleSetup event is sometimes before the component subscribes to it. Need to add a moduleSetup boolean check.

```
        if (this.oidcSecurityService.moduleSetup) {
            this.onModuleSetup();
        } else {
            this.oidcSecurityService.onModuleSetup.subscribe(() => {
                this.onModuleSetup();
            });
        }
```

**Comment by damienbod** [comment]
fixed in version 1.3.3

**Comment by bluebaroncanada** [comment]
Broken link.
